File: backend/src/main/java/com/vortexsynergy/backend/model/InventoryItem.java
package com.vortexsynergy.backend.model;

import com.vortexsynergy.backend.model.enums.InventoryStatus;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.time.Instant;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(
    name = "inventory_items",
    indexes = {
        @Index(name = "idx_inventory_receiver_status", columnList = "receiver_id, status"),
        @Index(name = "idx_inventory_resource", columnList = "resource_id")
    }
)
public class InventoryItem extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "receiver_id", nullable = false)
    private User receiver;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "resource_id", nullable = false)
    private Resource resource;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "claim_id")
    private Claim claim;

    @Column(nullable = false)
    private Integer quantityReceived;

    @Column(nullable = false)
    private Integer quantityAvailable;

    @Column(nullable = false)
    private Integer quantityConsumed;

    @Column(nullable = false)
    private Integer quantityExpired;

    @Column(length = 120)
    private String branchName;

    @Column(length = 255)
    private String storageLocation;

    private Instant stockedAt;

    private Instant lastConsumedAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private InventoryStatus status;

    @Column(length = 1000)
    private String notes;
}

File: mobile/lib/models/inventory_item.dart
class InventoryItem {
  const InventoryItem({
    required this.id,
    required this.resourceId,
    required this.claimId,
    required this.resourceTitle,
    required this.resourceType,
    required this.donorName,
    required this.quantityReceived,
    required this.quantityAvailable,
    required this.quantityConsumed,
    required this.quantityExpired,
    required this.unit,
    required this.status,
    required this.branchName,
    required this.storageLocation,
    required this.stockedAt,
    required this.lastConsumedAt,
    required this.foodExpiresAt,
    required this.medicineExpiryDate,
    required this.city,
    required this.area,
    required this.notes,
  });

  final String id;
  final String resourceId;
  final String? claimId;
  final String resourceTitle;
  final String resourceType;
  final String donorName;
  final int quantityReceived;
  final int quantityAvailable;
  final int quantityConsumed;
  final int quantityExpired;
  final String unit;
  final String status;
  final String? branchName;
  final String? storageLocation;
  final DateTime? stockedAt;
  final DateTime? lastConsumedAt;
  final DateTime? foodExpiresAt;
  final DateTime? medicineExpiryDate;
  final String city;
  final String area;
  final String? notes;

  factory InventoryItem.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic value) => value == null ? null : DateTime.tryParse(value.toString());

    return InventoryItem(
      id: json['id']?.toString() ?? '',
      resourceId: json['resourceId']?.toString() ?? '',
      claimId: json['claimId']?.toString(),
      resourceTitle: json['resourceTitle']?.toString() ?? '',
      resourceType: json['resourceType']?.toString() ?? '',
      donorName: json['donorName']?.toString() ?? '',
      quantityReceived: json['quantityReceived'] as int? ?? 0,
      quantityAvailable: json['quantityAvailable'] as int? ?? 0,
      quantityConsumed: json['quantityConsumed'] as int? ?? 0,
      quantityExpired: json['quantityExpired'] as int? ?? 0,
      unit: json['unit']?.toString() ?? '',
      status: json['status']?.toString() ?? '',
      branchName: json['branchName']?.toString(),
      storageLocation: json['storageLocation']?.toString(),
      stockedAt: parseDate(json['stockedAt']),
      lastConsumedAt: parseDate(json['lastConsumedAt']),
      foodExpiresAt: parseDate(json['foodExpiresAt']),
      medicineExpiryDate: parseDate(json['medicineExpiryDate']),
      city: json['city']?.toString() ?? '',
      area: json['area']?.toString() ?? '',
      notes: json['notes']?.toString(),
    );
  }
}
